src/V3Randomize.cpp

```diff
@@ -196,13 +207,121 @@ class RandomizeMarkVisitor final : public VNVisitor {
             staticRefp->classOrPackagep(VN_AS(staticRefp->varp()->user2p(), NodeModule));
         }
     }
+    // Mark nested MemberSel chain variables as globally constrained
+    void markNestedGlobalConstrainedRecurse(AstNode* nodep) {
+        UASSERT(nodep, "Node should not be null");
+
+        if (VN_IS(nodep, VarRef)) {
+            AstVar* const varp = VN_AS(nodep, VarRef)->varp();
+            if (varp->globalConstrained()) return;
+            varp->globalConstrained(true);
+        } else if (VN_IS(nodep, MemberSel)) {
+            AstMemberSel* const memberSelp = VN_AS(nodep, MemberSel);
+            if (memberSelp->varp()) {
+                AstVar* const varp = memberSelp->varp();
+                if (varp->globalConstrained()) return;
+                varp->globalConstrained(true);
+            }
+            markNestedGlobalConstrainedRecurse(memberSelp->fromp());
+        }
+    }
 
+    // Build MemberSel chain from variable path
+    AstNodeExpr* buildMemberSelChain(AstVarRef* rootVarRefp, const std::vector<AstVar*>& path) {
+        AstNodeExpr* exprp = rootVarRefp->cloneTree(false);
+        for (AstVar* memberVarp : path) {
+            AstMemberSel* memberSelp
+                = new AstMemberSel{rootVarRefp->fileline(), exprp, memberVarp};
+            memberSelp->user2p(m_classp);
+            exprp = memberSelp;
+        }
+        return exprp;
+    }
+
+    // Process a single constraint during nested constraint cloning
+    void processNestedConstraint(AstConstraint* const constrp, AstVarRef* rootVarRefp,
+                                 const std::vector<AstVar*>& newPath) {
+        UASSERT(constrp, "Constraint should not be null");
```

Comment:
`foreachMember` (where you call this) doesn't yield a nullptr, so this assert is a bit excessive. You can leave it if it makes you feel more comfortable, but a segfault here is just as good if you ever call this with nullptr.